yktracec/src/jitmodbuilder.cc

```diff
@@ -1912,6 +1922,54 @@ class JITModBuilder {
     finalise(AOTMod, &Builder);
     return JITMod;
   }
+
+  /// Handle promoted variables.
+  void handlePromote(CallInst *CI, MappableFrame *MPF, BasicBlock *BB,
+                     size_t CurBBIdx, size_t CurInstrIdx) {
+    // First lookup the constant value the trace is going to use.
+    uint64_t PConst = __yk_lookup_promote_usize();
+    Value *PConstLL =
+        Builder.getIntN(PointerSizedIntTy->getIntegerBitWidth(), PConst);
+    Value *PromoteVar = CI->getArgOperand(0);
+    Value *JITPromoteVar = getMappedValue(PromoteVar);
+
+    // Now emit a guard for the value, so that we deoptimise if the
+    // value ever changes.
+    MPF->LastSMCall = cast<CallInst>(CI->getNextNonDebugInstruction());
+
+    // We need a way to independently reference both the pre-promoted variable
+    // and the newly promoted constant in our mappings. We acheive this is a
```

Comment:
sp "achieve"

Reply:
Maybe something like:
```
Even though the call `y = yk_promote(x)` is not copied over into the trace and `y` is being replaced with a constant,  we need keep hold of the original value of `y` so it can be accessed during deoptimisation. The simplest way is by inserting a dummy instruction that will be copied into the trace, creating a mapping back to its AOT value.
```

This is not great either (this is difficult to describe), but maybe you can merge this with your comment somehow?

Reply:
I really struggled with this comment, and I'm not 100% sure I understand.

Let's recap. We have:

```
    Instruction *NewInst = SelectInst::Create(
        ConstantInt::get(Type::getInt1Ty(JITMod->getContext()), 0),
        JITPromoteVar, JITPromoteVar);
    Builder.Insert(NewInst);

    VMap[CI] = NewInst;
    insertAOTMap(CI, NewInst, CurBBIdx, CurInstrIdx);

    BasicBlock *FailBB = getGuardFailureBlock(BB, CurBBIdx, CI, CurInstrIdx);
    ...

    VMap[CI] = PConstLL;
```

You'd expect to be able to kill the dummy instruction, and use `JITPromoteVar` directly, but:
```
    VMap[CI] = JITPromoteVar;
    insertAOTMap(CI, NewInst, CurBBIdx, CurInstrIdx);
    ...
```

But this gives:
```
thread '<unnamed>' panicked at 'called `Option::unwrap()` on a `None` value', ykrt/src/frame/mod.rs:271:42
```

https://github.com/ykjit/yk/blob/9f26837c47a8c28ea913cd0d15c50bc7c7b04b12/ykrt/src/frame/mod.rs#L271

And around here I get lost. We try to look something up in the frame that the stackmap knows about, but t he frame doesn't...

@ptersilie remind me how it works again?

Reply:
Without the dummy instruction you map `JITPromoteVar` to two different AOT vars, which we can't handle. To fix this we could change the AOT mapping to allow multiple values, e.g. AOTMap[JITVar] = [AOT1, AOT2], but that's not a trivial change. Using a dummy instruction for now is much easier.

Reply:
Fixed in 32e3f65d6ee9735f1324f5b0fc46449af2be23bb

Reply:
New comment looks good!

Reply:
@ptersilie If you're happy with this please "resolve conversation".